### PR TITLE
System.Reflection: Replicate custom modifiers of method parameters in MemberRef signatures

### PR DESCRIPTION
**Pull request by stakx**
This is in response to https://github.com/dotnet/corefx/issues/29254.

`ModuleBuilder.GetMemberRefToken` reproduces a method's signature for use in a *MethodRef* metadata entry, but currently ignores custom modifiers placed on the method's parameters... which leads to non-matching signatures and `MissingMethodException`s when using e.g. reflection.

This commit adds a new method overload (in order to avoid a larger refactoring for now) that allows the inclusion of custom modifiers on parameters.

~~This is currently work in progress:~~

* [X] ~~Needs further testing.~~  
   I ran successful end-to-end tests using tests from Castle DynamicProxy, which has tests targeting this. I'm going to submit another PR with tests for dotnet/corefx.

* [X] ~~Are there any code paths into `ModuleBuilder.GetMemberRefToken` that are negatively affected by the change made?~~  
   It appears unlikely and did not find any (but it's pretty hard to keep track here, given all the branching and similarly named methods strewn across at least two classes).

* [X] ~~Are there any other code paths into the method that could profit from the change made?~~  
   Additional code paths have been included in the second commit of this PR.

* [X] ~~Shouldn't the same be done for the methods' return type (not just its parameters)?~~  
   Yes, but as it turns out, doing this for return parameters poses some unique problems that are perhaps better dealt with in a separate, dedicated PR.

### CODE REVIEW

**Comment by AtsushiKan** [comment]
One quick note: this will propagate custom modifiers attached to the parameter type itself but not any types are used to construct the type. E.g. a parameter of type `int[]` can have a custom modifier attached to the `int[]` or the `int` - you'll only propagate the former.

This is a long standing defect in how Reflection thinks about custom modifiers and it's not likely you could address this here without majorly impacting the cost of the fix. Just noting it here so we know the limitation exists, that this fix still solves the problem you wanted to solve with that limitation and that we're proceeding with knowledge of that.

**Comment by stakx** [comment]
@AtsushiKan - I've suggested above that perhaps this PR shouldn't just focus on the `ldtoken` code path and on parameter cmods. Perhaps, for consistency, other code paths (e.g. `call` / `calli`) and return parameter cmods should also be taken into account. That would be a larger refactoring involving a variety of things, e.g.:

* Changing internal methods to accept additional parameters (`Type[][] returnType[Required|Optional]CustomModifiers` and `parameterType[Required|Optional]CustomModifiers`).
* Turning methods like `MethodBuilder.GetMethodBaseReturnType` into `MethodBuilder.GetMethodBaseReturnParameter` so that cmods can be retrieved.
* Some additional array allocations due to having to query `ParameterInfo.Get[Required|Optional]CustomModifiers()` in more places.
* Adding a new public API method overload `ILGenerator.EmitCalli` that also has parameters for cmods.

Is this the right time and place to do all of this (well, maybe not the last bullet point)? Or should I stick to the original, narrow scenario that led to this PR for now?

File: src/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs [hunk 1]
```diff
@@ -484,15 +488,24 @@ private int GetMemberRefToken(MethodBase method, IEnumerable<Type> optionalParam
...
 
+            Debug.Assert(parameterTypeRequiredCustomModifiers == null
+                     || (parameterTypeRequiredCustomModifiers.Length == (parameterTypes?.Length ?? 0)));
+
+            Debug.Assert(parameterTypeOptionalCustomModifiers == null
+                     || (parameterTypeOptionalCustomModifiers.Length == (parameterTypes?.Length ?? 0)));
+
             if (parameterTypes != null)
             {
+                int i = 0;
```
Comment (AtsushiKan) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs]: I would just turn this into a `for (int i = …)` loop.

**Comment by stakx** [comment]
@AtsushiKan - I started working on return parameters, but found that this poses some additional problems and puzzles to solve. I think it'd be better to deal with return parameters in a separate, dedicated PR. In my opinion, the present PR is ready for review / merging.

<details>
<summary>(Expand this to see an example of an additional problem unique to return parameters.)</summary>
<br>

In this PR, often a `Type parameterType` gets replaced with a `ParameterInfo parameter` from which custom modifiers can be accessed. If we follow the same procedure for return parameters, we need to e.g. deal with this existing method:

https://github.com/dotnet/coreclr/blob/e353a220a673fa54e98f53c4dcc5793badd4d69e/src/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs#L231-L232

Either a corresponding new method `ParameterInfo RuntimeConstructorInfo.GetReturnParameter()` is added, or the above existing method is removed (perhaps preferable as it doesn't appear to make too much sense to begin with).

It's these kinds of decisions I'd like to shift to a separate PR to keep the present one focused.
</details>

File: src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs [hunk 2]
```diff
@@ -191,6 +191,56 @@ internal static SignatureHelper GetTypeSigToken(Module module, Type type)
...
+        {
+            // This method tries to allocate as few arrays as possible. In the absence of parameters or
+            // custom modifiers, arrays or single items will default to null. The returned arrays will
+            // generally be fed back to a `SignatureHelper` which must perform null checks where necessary.
+
+            Type[] types = null;
+            Type[][] requiredCustomModifiers = null;
+            Type[][] optionalCustomModifiers = null;
+
+            if (parameters?.Length > 0)
```
Comment (AtsushiKan) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs]: Every caller of this passes a non-null `parameters`. The `?` isn't needed.

File: src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs [hunk 3]
```diff
@@ -191,6 +191,56 @@ internal static SignatureHelper GetTypeSigToken(Module module, Type type)
...
+            {
+                types = new Type[parameters.Length];
+                for (int i = 0; i < parameters.Length; i++)
+                {
+                    Debug.Assert(parameters[i] != null);
+
+                    types[i] = parameters[i].ParameterType;
+
+                    Type[] rcms = parameters[i].GetRequiredCustomModifiers();
+                    if (rcms?.Length > 0)
```
Comment (AtsushiKan) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs]: The `?` isn't needed. `GetRequiredCustomModifiers` never returns a null array.

File: src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs [hunk 4]
```diff
@@ -191,6 +191,56 @@ internal static SignatureHelper GetTypeSigToken(Module module, Type type)
...
+                    {
+                        if (requiredCustomModifiers == null)
+                        {
+                            requiredCustomModifiers = new Type[parameters.Length][];
+                        }
+                        requiredCustomModifiers[i] = rcms;
+                    }
+
+                    Type[] ocms = parameters[i].GetOptionalCustomModifiers();
+                    if (ocms?.Length > 0)
```
Comment (AtsushiKan) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs]: Same comment regarding `?`

File: src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs (hunk 2 above)
Comment (stakx) [reply src/System.Private.CoreLib/src/System/Reflection/Emit/SignatureHelper.cs]: Thanks for the review. I've replaced all three null-conditional member accesses `?.` with `.`.

**Comment by TAGC** [comment]
Thanks for the work @stakx. 👍

**Comment by stakx** [comment]
@TAGC - Thanks for the thumbs-up, I appreciate it! Not quite done yet, but we're getting there.

**Pull request merged**